mpc-core/src/protocols/rep3/yao/evaluator.rs

```diff
@@ -0,0 +1,274 @@
+        let block = self.receive_block_from(PartyID::ID1)?;
+        self.hash.update(block.as_ref()); // "Receive" from ID2
+
+        Ok(block)
+    }
+
+    /// Read `n` `Block`s from the channel.
+    #[inline(always)]
+    fn read_blocks(&mut self, n: usize) -> IoResult<Vec<Block>> {
+        (0..n).map(|_| self.receive_block()).collect()
+    }
+
+    /// Read a Wire from the reader.
+    pub fn read_wire(&mut self) -> IoResult<WireMod2> {
+        let block = self.receive_block()?;
+        Ok(WireMod2::from_block(block, 2))
+    }
+
+    /// Receive a bundle of wires over the established channel.
+    pub fn receive_bundle(&mut self, n: usize) -> IoResult<BinaryBundle<WireMod2>> {
+        let mut wires = Vec::with_capacity(n);
+        for _ in 0..n {
+            let wire = WireMod2::from_block(self.receive_block()?, 2);
+            wires.push(wire);
+        }
+
+        Ok(BinaryBundle::new(wires))
+    }
+
+    /// Evaluates an 'and' gate given two inputs wires and two half-gates from the garbler.
+    ///
+    /// Outputs C = A & B
+    ///
+    /// Used internally as a subroutine to implement 'and' gates for `FancyBinary`.
+    fn evaluate_and_gate(
+        &mut self,
+        a: &WireMod2,
+        b: &WireMod2,
+        gate0: &Block,
+        gate1: &Block,
+    ) -> WireMod2 {
+        let gate_num = self.current_gate();
+        let g = tweak2(gate_num as u64, 0);
+
+        let [hash_a, hash_b] = hash_wires([a, b], g);
+
+        // garbler's half gate
+        let l = WireMod2::from_block(
+            Block::conditional_select(&hash_a, &(hash_a ^ *gate0), (a.color() as u8).into()),
+            2,
+        );
+
+        // evaluator's half gate
+        let r = WireMod2::from_block(
+            Block::conditional_select(&hash_b, &(hash_b ^ *gate1), (b.color() as u8).into()),
+            2,
+        );
+
+        l.plus_mov(&r.plus_mov(&a.cmul(b.color())))
+    }
+}
+
+impl<'a, N: Rep3Network> Fancy for Rep3Evaluator<'a, N> {
+    type Item = WireMod2;
+    type Error = EvaluatorError;
+
+    fn constant(&mut self, _: u16, _q: u16) -> Result<WireMod2, EvaluatorError> {
+        Ok(self.read_wire()?)
+    }
+
+    fn output(&mut self, x: &WireMod2) -> Result<Option<u16>, EvaluatorError> {
+        let q = 2;
+        let i = self.current_output();
+
+        // Receive the output ciphertext from the garbler
+        let ct = self.read_blocks(q as usize)?;
+
+        // Attempt to brute force x using the output ciphertext
+        let mut decoded = None;
+        for k in 0..q {
+            let hashed_wire = x.hash(output_tweak(i, k));
+            if hashed_wire == ct[k as usize] {
+                decoded = Some(k);
+                break;
+            }
+        }
+
+        if let Some(output) = decoded {
+            Ok(Some(output))
+        } else {
+            Err(EvaluatorError::DecodingFailed)
+        }
+    }
+}
+
+impl<'a, N: Rep3Network> FancyBinary for Rep3Evaluator<'a, N> {
```

Comment:
I think we might want a variant that buffers the full circuit and sends it in one go.

Reply:
Lets talk tomorrow about that, whether I add a new garbler file or just rewrite it to your variant

Reply:
Should be easy to add a second file, the core logic is in GCUtils

Reply:
added both versions